### DISCUSSION TITLE
Diagnostic a pathological behavior: solution?

### DISCUSSION THREAD

**Discussion by jecampagne** [Joshuaalbert/jaxns#28] [Q&A]
Dear experts,

I have posted this message to `Numpyro` forum, and they invite me to ask the question here as their  `NestedSampler` class is a wrapper to `JaxNS` ([see here](http://num.pyro.ai/en/stable/contrib.html?highlight=Nested#numpyro.contrib.nested_sampling.NestedSampler.run)) . Also I must confess that I 'm not an experienced user of Nested SAmpler so my question may be  not well formulated. 

So, I have a `model` which I sample with the `NUTS` sampler. Schematically, I do the following (sorry this is in the Numpyro language)
1) define my model using `numpyro.sample` statements to define priors (which are for this exemple all Gaussian distributions) and likelihood
2) I use  `fix_cond_model = numpyro.handlers.condition(model, <parmeters defult values>` to generate some `data` thanks to

```python
tr = numpyro.handlers.trace(seed(fix_cond_model, rng_key))
res = tr.get_trace()
data = res['data']['value']
```
Then,  I proceed to the MCMC run et finally get the samples:

```python
cond_model = numpyro.handlers.condition(model, {'data':data})
nuts_kernel = numpyro.infer.NUTS(cond_model,
                                 init_strategy=numpyro.infer.init_to_sample(),
                                 max_tree_depth=2) # max_tree_depth=10

rng_key = jax.random.PRNGKey(0)

mcmc = numpyro.infer.MCMC(nuts_kernel, 
                          num_warmup=1000, 
                          num_samples=10000,
                          num_chains=1,
                          jit_model_args=True,
                          progress_bar=True)

mcmc.run(rng_key, extra_fields=('potential_energy',))
samples= mcmc.get_samples()
```

So far so good, now I wander if I can use the `NestedSampler`? I have tried
```python

cond_model = numpyro.handlers.condition(model, {'data':data})
ns = NestedSampler(cond_model)
ns.run(jax.random.PRNGKey(42))
samples= ns.get_samples(jax.random.PRNGKey(3),num_samples=10_000)
```

From the function calls point of view, it seems ok from `Numpyro` developers, but the sampling of the variables is clearly pathologic, 

![image](https://user-images.githubusercontent.com/20539759/137580168-f6b94750-c415-4393-9c83-e02357258beb.png)

And here are the arviz kde plots
![image](https://user-images.githubusercontent.com/20539759/137579999-0594d432-31c7-4639-b29a-f835579c1a4f.png)

I certainly miss something. Any idea are welcome. Thanks

**Comment by jecampagne** [comment]
Dear @Joshuaalbert,
If you prefer that I implement my model directly in JaxNS, I am ready to do so but I need some help to transpose Numpyro model which concern 20 variable priors and a MultovariateGaussian likelihood, as well as a piece of code to generate observations from this modelling. 
Thanks

**Reply by Joshuaalbert** [reply]
I'd be happy to help. Firstly, can you downgrade jax to 0.0.6 for these tests, until 1.0 is released?

**Reply by jecampagne** [reply]
Heuu it is jaxns that I should downgrade, isn't it?

**Comment by jecampagne** [comment]
Hello @Joshuaalbert, 
I try top convert the Numpyro model in JaxNS (amy be I am close). 

Here is my Nympyro code

```python
def model():
    #  Cosmological params (test13 JEC)
    Omega_c = numpyro.sample('Omega_c', dist.Uniform(0.1, 0.4))
    sigma8 = numpyro.sample('sigma8', dist.Uniform(0.5, 1.2))
    Omega_b = numpyro.sample('Omega_b', dist.Uniform(0.01, 0.09))
    h = numpyro.sample('h', dist.Uniform(0.4, 1.0))
    n_s = numpyro.sample('n_s', dist.Uniform(0.5, 1.25))
    w0 = numpyro.sample('w0', dist.Uniform(-2.0, -0.0001))
    # Astrophysical params
    A = numpyro.sample('A', dist.Uniform(0., 2.5))
    eta = numpyro.sample('eta', dist.Uniform(0., 6.))

    
    # parameters for systematics
    m = [numpyro.sample('m%d'%i, dist.Normal(0.012, 0.023)) 
         for i in range(1,5)]
    dz1 = numpyro.sample('dz1', dist.Normal(0.001, 0.016)) 
    dz2 = numpyro.sample('dz2', dist.Normal(-0.019, 0.013)) 
    dz3 = numpyro.sample('dz3', dist.Normal(0.009, 0.011)) 
    dz4 = numpyro.sample('dz4', dist.Normal(-0.018, 0.022)) 
    
    
    # Now that params are defined, here is the forward model
    cosmo = FiducialCosmo(Omega_c=Omega_c, sigma8=sigma8, Omega_b=Omega_b,
                          h=h, n_s=n_s, w0=w0)
    
    signal = model_fn(get_params_vec(cosmo, m, [dz1, dz2, dz3, dz4], [A, eta]))
    
    # And here we define the likelihood
    numpyro.sample('cl_wl', dist.MultivariateNormal(signal, C), obs=cl_obs)
```
Here is my tentative to setup a JaxNS code:

```python
#@jax.jit
#code from https://github.com/google/jax/issues/2314
def multi_gauss_logpdf(x, mean, cov):
    """ Calculate the probability density of a
        sample from the multivariate normal. """
    D = mean.shape[0]
    (sign, logdet) = np.linalg.slogdet(cov)
    p1 = D*np.log(2*np.pi) + logdet
    p2 = (x-mean).T @ np.linalg.inv(cov) @ (x-mean)
    return -1./2 * (p1 + p2)



def solve(cov,cl_obs):
    def log_lik(Omega_c, sigma8, Omega_b,h, n_s, w0, A, eta,
                m1, m2,m3, m4,
                dz1, dz2, dz3, dz4,
                **kwargs):
        
        cosmo = FiducialCosmo(Omega_c=Omega_c, sigma8=sigma8, Omega_b=Omega_b,
                          h=h, n_s=n_s, w0=w0)
        
        signal = model_fn(get_params_vec(cosmo, [m1, m2,m3, m4], [dz1, dz2, dz3, dz4], [A, eta]))
        
        
        
        ### Comment faire numpyro.sample('cl_wl', dist.MultivariateNormal(signal, cov), obs='cl_obs')
        return multi_gauss_logpdf(cl_obs,signal,cov)
  
    prior_chain = PriorChain(UniformPrior('Omega_c',0.1, 0.4),
                                UniformPrior('sigma8',0.5, 1.2),
                                UniformPrior('Omega_b',0.01, 0.09),
                                UniformPrior('h',0.4, 1.0),
                                UniformPrior('n_s',0.5, 1.25),
                                UniformPrior('w0',-2.0, -0.0001),
                                UniformPrior('A',0.,2.5),
                                UniformPrior('eta',0.,6.),
                                NormalPrior('m1',0.012, 0.023),
                                NormalPrior('m2',0.012, 0.023),
                                NormalPrior('m3',0.012, 0.023),
                                NormalPrior('m4',0.012, 0.023),
                                NormalPrior('dz1',0.001, 0.016),
                                NormalPrior('dz2',-0.019, 0.013),
                                NormalPrior('dz3',0.009, 0.011),
                                NormalPrior('dz4',-0.018, 0.022)
                            )
    
    print('num_live_points:',prior_chain.U_ndims*500)
                             
    ns = jaxns.nested_sampling.NestedSampler(log_lik, prior_chain,
                       num_live_points=prior_chain.U_ndims*500)
    
    print('Go...')
                             
    results = ns(jax.random.PRNGKey(32564))
    
    return results

## Go.

results = solve(C,data)

```
Now I try to run this code on GPU...
num_live_points: 8000
Go...
```

But I get this crash trace back

```
2021-10-20 15:11:15.189842: W external/org_tensorflow/tensorflow/core/common_runtime/bfc_allocator.cc:461] Allocator (GPU_0_bfc) ran out of memory trying to allocate 933.17GiB (rounded to 1001989427968)requested by op 
2021-10-20 15:11:15.190098: W external/org_tensorflow/tensorflow/core/common_runtime/bfc_allocator.cc:472] *___________________________________________________________________________________________________
2021-10-20 15:11:15.190153: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2085] Execution of replica 0 failed: RESOURCE_EXHAUSTED: Out of memory while trying to allocate 1001989427840 bytes.
Traceback (most recent call last):
  File "DESY_shear_wl_JaxNS_test_GPU.py", line 524, in <module>
    results = solve(C,data)
  File "DESY_shear_wl_JaxNS_test_GPU.py", line 516, in solve
    results = ns(jax.random.PRNGKey(32564))
  File "python3.8/site-packages/jaxns/nested_sampling.py", line 616, in __call__
    state = self.initial_state(key)
  File "python3.8/site-packages/jaxns/nested_sampling.py", line 225, in initial_state
    live_points_U, live_points_X, log_L_live = vmap(single_sample)(
  File "python3.8/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "python3.8/site-packages/jax/_src/api.py", line 1289, in batched_fun
    out_flat = batching.batch(
  File "/python3.8/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "python3.8/site-packages/jaxns/nested_sampling.py", line 208, in single_sample
    log_L = self.loglikelihood(**X)
  File "/python3.8/site-packages/jaxns/nested_sampling.py", line 148, in corrected_likelihood
    log_L = loglikelihood(**x)
  File "DESY_shear_wl_JaxNS_test_GPU.py", line 484, in log_lik
    signal = model_fn(get_params_vec(cosmo, [m1, m2,m3, m4], [dz1, dz2, dz3, dz4], [A, eta]))
  File "python3.8/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "python3.8/site-packages/jax/_src/api.py", line 411, in cache_miss
    out_flat = xla.xla_call(
  File "python3.8/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "python3.8/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File python3.8/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "python3.8/site-packages/jax/interpreters/batching.py", line 177, in process_call
    vals_out = call_primitive.bind(f, *vals, **params)
  File "python3.8/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/python3.8/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/python3.8/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "python3.8/site-packages/jax/core.py", line 615, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/python3.8/site-packages/jax/interpreters/xla.py", line 625, in _xla_call_impl
    out = compiled_fun(*args)
  File "/python3.8/site-packages/jax/interpreters/xla.py", line 960, in _execute_compiled
    out_bufs = compiled.execute(input_bufs)
jax._src.traceback_util.UnfilteredStackTrace: RuntimeError: RESOURCE_EXHAUSTED: Out of memory while trying to allocate 1001989427840 bytes.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "DESY_shear_wl_JaxNS_test_GPU.py", line 524, in <module>
    results = solve(C,data)
  File "DESY_shear_wl_JaxNS_test_GPU.py", line 516, in solve
    results = ns(jax.random.PRNGKey(32564))
  File "python3.8/site-packages/jaxns/nested_sampling.py", line 616, in __call__
    state = self.initial_state(key)
  File "/python3.8/site-packages/jaxns/nested_sampling.py", line 225, in initial_state
    live_points_U, live_points_X, log_L_live = vmap(single_sample)(
  File "/python3.8/site-packages/jaxns/nested_sampling.py", line 208, in single_sample
    log_L = self.loglikelihood(**X)
  File "python3.8/site-packages/jaxns/nested_sampling.py", line 148, in corrected_likelihood
    log_L = loglikelihood(**x)
  File "DESY_shear_wl_JaxNS_test_GPU.py", line 484, in log_lik
    signal = model_fn(get_params_vec(cosmo, [m1, m2,m3, m4], [dz1, dz2, dz3, dz4], [A, eta]))
RuntimeError: RESOURCE_EXHAUSTED: Out of memory while trying to allocate 1001989427840 bytes.
```

Notice that my cov matrix is 500x500 and the cl_obs is a vector 1D of size 500. With Numpyro JAXNS wrapper I had no crash (but not understandable results too).

Any help is welcome. Thanks

PS note added 21-oct-2021-10:35 ParisTime: Du Phan from Numpyro told me that an alternative to `multi_gauss_logpdf(x, mean, cov)`is to use `dist.MultivariateNormal(loc, cov).log_prob(x)`. But this does not change the pathological behavior of the results.

**Comment by jecampagne** [comment]
The memory load clearly depend on the # of live points, as setting to 1,000 get a 100GiB order (that crashes too) so, I down grade to num live pts 100. But what is strange, that is the Numpyro implementation runs with 10,000 live points
```python
ns = NestedSampler(cond_model, num_slices=10,
                   num_live_points=10000)
ns.run(jax.random.PRNGKey(42))
``` 
on the same kind of GPU hardware.  So I do not understand what is going wrong in my native JaxNS calls...


With 100 num live points, the job runs up to the end, but the results are very poor, worse than the Numpyro implementation results.
![image](https://user-images.githubusercontent.com/20539759/138117529-170c47b8-88e2-4323-a69a-100dc9d13b1c.png)

By the way a NUTS run with (Normal priors instead of Uniform priors) gives this kind of results to give you an idea
![image](https://user-images.githubusercontent.com/20539759/138117910-d6f76c7a-c615-4707-b852-e92e67fa9649.png)

**Reply by jecampagne** [reply]
Hi @Joshuaalbert 

I just discover that for some parameter sets, it results NaN  likelihood value  . So, I guess these NaN are making the mess in your algorithm. If it is the case, it would be good to rise an exception, no?

**Reply by Joshuaalbert** [reply]
@jecampagne can you pull the latest and try again. The API has changed slightly, see the tutorial here: https://github.com/Joshuaalbert/jaxns/blob/master/examples/simple/mvn_data_mvn_prior.ipynb

**Comment by Joshuaalbert** [comment]
@jecampagne does this example require any data? If not, I could add it to the examples section.

**Reply by jecampagne** [reply]
Ho, I will let you know but currently I have an other hot topic to finish. Sorry.
